### ISSUE TITLE
[1.20 regression] False positive for checking reachability on dict with single-variant enum keys

### ISSUE THREAD

**Issue by kmurphy4** [python/mypy#21409]
**Bug Report**
After upgrading from 1.19.x to 1.20.x, I'm seeing a false positive with reachability checking for `dict`s with single-variant `Enum` keys.

**To Reproduce**

https://mypy-play.net/?mypy=latest&python=3.12&flags=warn-unreachable&gist=3ba2a6f89ea69ebf0ce78f0d91fb0e7f

```py
import enum

class Color(enum.Enum):
    RED = enum.auto()

def foo(colors: list[Color]) -> None:
    counts: dict[Color, int] = {}
    for color in colors:
        if color not in counts:
            counts[color] = 0  # error: Statement is unreachable  [unreachable]
        counts[color] += 1
```

If I [add a second `Enum` variant](https://mypy-play.net/?mypy=1.20.2&python=3.12&flags=warn-unreachable&gist=3ab07115fc955b11e27bf639c337cd8c), the error goes away.

Interestingly, the error also goes away if I [remove the explicit type annotation](https://mypy-play.net/?mypy=1.20.2&python=3.12&flags=warn-unreachable&gist=8fad802fc75f8f5db0dae53f1a297352).

**Expected Behavior**

[With v1.19.1](https://mypy-play.net/?mypy=1.19.1&python=3.12&flags=warn-unreachable&gist=3ba2a6f89ea69ebf0ce78f0d91fb0e7f), I get no errors.

**Comment by hauntsaninja** [comment]
Thanks, fixed in #21411
